File: src/StreamDb/Internal/DbStructure/IndexPage.cs
using System;
using System.IO;
using JetBrains.Annotations;
using StreamDb.Internal.Support;

namespace StreamDb.Internal.DbStructure
{
    /// <summary>
    /// Content of a single index page
    /// </summary>
    public class IndexPage : IStreamSerialisable
    {

        const int EntryCount = 126; // 2+4+8+16+32+64
        const int PackedSize = 3276; // (16+5+5) * 126
        
        /// <summary> This is the implicit root index. It is not allowed as a real document ID </summary>
        public static readonly Guid NeutralDocId = new Guid(new byte[] { 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127 });
        /// <summary> This is an ID that means 'no document'. It is not allowed as a real document ID. </summary>
        public static readonly Guid ZeroDocId = Guid.Empty;


        [NotNull, ItemNotNull] private readonly VersionedLink[] _links;
        [NotNull] private readonly Guid[] _docIds;

        /*

            Layout: [ Doc Guid (16 bytes) | PageLink[0] (5 bytes) | PageLink[1] (5 bytes) ] --> 26 bytes
            We can fit 157 in a 4k page. Gives us 6 ranks (126 entries) -> 3276 bytes
            Our pages are currently 4061 bytes, so we have plenty of spare space if we can find useful metadata to store.

            We assume but don't store a root page with guid {127,127...,127}. The first two entries are 'left' and 'right' on the second level.

        */

        public IndexPage()
        {
            _links = new VersionedLink[EntryCount];
            for (int i = 0; i < EntryCount; i++) { _links[i] = new VersionedLink(); }

            _docIds = new Guid[EntryCount];
        }

        const int SAME =  0;
        const int LESS =  -1;
        const int GREATER =  1;

        /// <summary>
        /// Try to add a new link to the index.
        /// Returns true if written, false if the index page has no space for this entry.
        /// If the document already exists, an exception will be thrown.
        /// </summary>
        /// <param name="docId">Unique ID of the document to be inserted</param>
        /// <param name="pageId">PageID of the LAST page in the document's chain.</param>
        /// <returns>True if written, false if not</returns>
        public bool TryInsert(Guid docId, int pageId)
        {
            var index = Find(docId);
            if (index < 0 || index >= EntryCount) return false; // no space

            if (_docIds[index] != ZeroDocId) throw new Exception("Tried to insert a duplicate document ID");

            // found a space. Stick it in.
            _links[index].WriteNewLink(pageId, out _);
            _docIds[index] = docId;
            return true;

        }

        /// <summary>
        /// Try to find a link in this index page. Returns true if found, false if not found.
        /// If found, this will return up to two page options. Use the newest one with a valid CRC in the page.
        /// </summary>
        /// <param name="docId">Document ID to find</param>
        /// <param name="link">If found, this is the page link options. May be null</param>
        public bool Search(Guid docId, out VersionedLink link) {
            link = null;

            var index = Find(docId);
            if (index < 0 || index >= EntryCount) return false; // not found
            if (_docIds[index] == ZeroDocId) return false; // not found
            if (_docIds[index] != docId) throw new Exception("IndexPage.Search: Logic error");

            link = _links[index];

            return true;
        }

        /// <summary>
        /// Update a link with a new PageID. The oldest link will be updated.
        /// Returns true if a change was made. False if the link was not found in this index page
        /// </summary>
        /// <param name="docId">ID of document to update</param>
        /// <param name="pageId">PageID of the LAST page in the new document chain to be inserted</param>
        /// <param name="expiredPage">If an old value is lost, this is PageID. Otherwise -1</param>
        /// <remarks>If an existing chain is de-linked by this, all the pages should be added to the free list</remarks>
        public bool Update(Guid docId, int pageId, out int expiredPage) {
            expiredPage = -1;

            // find the entry to update
            var index = Find(docId);
            if (index < 0 || index >= EntryCount) return false; // not found
            if (_docIds[index] == ZeroDocId) return false; // not found
            if (_docIds[index] != docId) throw new Exception("IndexPage.Search: Logic error");

            _links[index].WriteNewLink(pageId, out expiredPage);
            return true;
        }

        /// <summary>
        /// Find tries to find an entry index by a guid key. This is used in insert, search, update.
        /// If no such entry exists, but there is a space for it, you will get a valid index whose
        /// `_docIds` entry is Guid.Zero -- so always check.
        /// </summary>
        private int Find(Guid target) {
            // the implicit node:
            var cmpNode = NeutralDocId;
            int leftIdx = 0;
            int rightIdx = 1;

            var current = -1;

            // loop start
            for (int i = 0; i < 7; i++)
            {
                switch (cmpNode.CompareTo(target))
                {
                    case SAME: return current;

                    case LESS:
                        // move left
                        current = leftIdx;
                        break;

                    case GREATER:
                        // move right
                        current = rightIdx;
                        break;

                    default: throw new Exception("IndexTree.TryInsert: Unexpected case.");
                }

                // update next step pointers
                leftIdx = (current * 2) + 2;
                rightIdx = (current * 2) + 3;

                // check we're in bounds
                if (current < 0) throw new Exception("IndexTree.TryInsert: Logic error");
                if (current >= EntryCount) return -1;
                
                cmpNode = _docIds[current];
                if (cmpNode == ZeroDocId) { return current; } // empty space
            }
            // loop end

            throw new Exception("IndexTree.TryInsert: Out of loops bounds. Exited due to safety check.");
        }

        /// <inheritdoc />
        public void Defrost(Stream source)
        {
            if (source == null || source.Length < PackedSize) throw new Exception("IndexPage.FromBytes: data was too short.");
            var r = new BinaryReader(source);

            for (int i = 0; i < EntryCount; i++)
            {
                var bytes = r.ReadBytes(16);
                if (bytes == null) throw new Exception("Failed to read doc guid");
                _docIds[i] = new Guid(bytes);


                _links[i].Defrost(r.BaseStream);
            }
        }

        /// <inheritdoc />
        public Stream Freeze()
        {
            var ms = new MemoryStream(PackedSize);
            var w = new BinaryWriter(ms);

            for (int i = 0; i < EntryCount; i++)
            {
                w.Write(_docIds[i].ToByteArray());
                _links[i].Freeze().CopyTo(ms);
            }

            ms.Seek(0, SeekOrigin.Begin);
            return ms;
        }
    }
}

File: golang/internal/structure/index_page.go
package structure

import (
	. "golang/internal/comparable"
	"golang/internal/support"
	"bytes"
	"errors"
	"io"
)


const (
	IndexEntryCount = 126 // 2+4+8+16+32+64
	IndexPackedSize = 3279 // (16+5+5) * 126

	same = 0
	less = -1
	greater = 1
)

var (
	zeroDocId = support.NewZeroId()
	neutralDocId = support.NewNeutralId()
)

// IndexPage holds the content of a single index page.
// The indexing system is a chained list of separate binary trees.
/*

   Layout: [ Doc Guid (16 bytes) | PageLink[0] (5 bytes) | PageLink[1] (5 bytes) ] --> 26 bytes
   We can fit 157 in a 4k page. Gives us 6 ranks (126 entries) -> 3276 bytes
   Our pages are currently 4061 bytes, so we have plenty of spare space if we can find useful metadata to store.

   We assume but DON'T store a root page with guid {127,127...,127}. The first two entries are 'left' and 'right' on the second level.

*/
type IndexPage struct {
	links []*support.VersionedLink
	docIds []*support.SerialId
}

func NewIndexPage() *IndexPage{
	page := IndexPage{
		links:  make([]*support.VersionedLink, IndexEntryCount),
		docIds: make([]*support.SerialId, IndexEntryCount),
	}
	for i := 0; i < IndexEntryCount; i++ {
		page.links[i] = support.NewVersionedLink()
		page.docIds[i] = support.NewZeroId()
	}

	return &page
}

func (page *IndexPage)TryInsert(docId *support.SerialId, pageId int) (bool,error) {
	index := find(page, docId)
	if index < 0 || index > IndexEntryCount {return false, nil} // no space left

	if Is(page.docIds[index]).NotEqual(zeroDocId) {return false, errors.New("tried to insert a duplicate document id")}

	// ok, we have a space
	_ = page.links[index].WriteNewLink(pageId)
	page.docIds[index] = docId
	return true, nil
}

// Search tries to find a link in this index page. Returns true if found, false if not found.
// If found, this will return up to two page-options. Use the newest one with a valid CRC in the page.
func (page *IndexPage)Search(docId *support.SerialId) (link *support.VersionedLink, found bool){
	index := find(page, docId)
	if index < 0 || index >= IndexEntryCount {return nil, false}
	if Is(page.docIds[index]).EqualTo(zeroDocId) {return nil, false}
	if Is(page.docIds[index]).NotEqual(docId) {panic("index page Search: logic error")}

	return page.links[index], true
}

// Update updates a link with a new PageID. The oldest link will be updated.
// Returns true if a change was made. False if the link was not found in this index page.
func (page *IndexPage)Update(docId *support.SerialId, pageId int) (expiredPage int, found bool){
	index := find(page, docId)
	if index < 0 || index >= IndexEntryCount {return -1, false}
	if Is(page.docIds[index]).EqualTo(zeroDocId) {return -1, false}
	if Is(page.docIds[index]).NotEqual(docId) {panic("index page Update: logic error")}

	expiredPage = page.links[index].WriteNewLink(pageId)
	return expiredPage, true
}

// Remove updates a link to set an invalid link. Both versions of the existing link will be lost.
// Returns true if a change was made. False if the link was not found in this index page
func(page *IndexPage) Remove(docId *support.SerialId) bool{
	index := find(page, docId)
	if index < 0 || index >= IndexEntryCount {return false}
	if Is(page.docIds[index]).EqualTo(zeroDocId) {return false}
	if Is(page.docIds[index]).NotEqual(docId) {panic("index page Remove: logic error")}

	page.links[index] = support.NewVersionedLink() // reset
	return true
}

// Freeze converts to a byte stream
func(page *IndexPage) Freeze() support.LengthReader {
	buf := bytes.NewBuffer([]byte{})

	for i := 0; i < IndexEntryCount; i++ {
		// Alternate doc-ids and page-version-links (both fixed size)
		id := page.docIds[i].Freeze()
		_, _ = io.Copy(buf, id)

		link := page.links[i].Freeze()
		_, _ = io.Copy(buf, link)
	}

	return buf
}

// Defrost populates data from a byte stream
func(page *IndexPage) Defrost(reader io.Reader) error{
	for i := 0; i < IndexEntryCount; i++ {
		id := support.NewZeroId()
		err := id.Defrost(reader)
		if err != nil {return err}
		page.docIds[i] = id

		link := support.NewVersionedLink()
		err = link.Defrost(reader)
		if err != nil {return err}
		page.links[i] = link
	}
	return nil
}

// find tries to find an entry index by a guid key. This is used in insert, search, update.
// If no such entry exists, but there is a space for it, you will get a valid index whose
// `docIds` entry is Guid.Zero -- so always check.
func find(page *IndexPage, target *support.SerialId) int {
	// the implicit node:
	cmpNode := neutralDocId
	leftIdx := 0
	rightIdx := 1

	current := -1

	// step down through the implicit binary tree
	for i := 0; i < 7; i++ {
		switch cmpNode.CompareTo(target) {
		case same: return current // found it!
		case less:
			// move left
			current = leftIdx

		case greater:
			// move right
			current = rightIdx

		default:
			panic("comparable returned unexpected value")
		}

		// update 'next' pointers
		leftIdx = (current*2) + 2
		rightIdx = (current*2) + 3

		// make sure we're still in bounds
		if current < 0 {panic("index tree TryInsert: Logic error")}
		if current >= IndexEntryCount {return -1}

		cmpNode = page.docIds[current]
		if Is(cmpNode).EqualTo(zeroDocId) {return current}
	}
	panic("index tree TryInsert: out of loop bounds")
}